### PR TITLE
[Raiden Weight Sync 4/7] Target-free and streaming weight conversion for vLLM

### PR DESCRIPTION
**Pull request by YixuanWang-99**
## Overview
Part of the stacked Raiden weight-sync enablement PR chain replacing #5089.

**Stack:**
- [M7] https://github.com/AI-Hypercomputer/maxtext/pull/5166: Cross-repo drift guard
- [M1] https://github.com/AI-Hypercomputer/maxtext/pull/5167: Inhomogeneous layer cycle support in raiden_unscan
- [M2] https://github.com/AI-Hypercomputer/maxtext/pull/5168: MoE 128-lane layout and padding (pairs with Tunix T5)
- **[M3] AI-Hypercomputer/maxtext (this PR)**: Target-free streaming weight conversion
- [M5] AI-Hypercomputer/maxtext: Shared expert gate weight_dtype fix
- [M4] AI-Hypercomputer/maxtext: TrainingEngine Raiden FFI integration (pairs with Tunix T2a)
- [M6] AI-Hypercomputer/maxtext: Rollout cleanup

## Details
- Implements `TargetFreeWeightConverter` to convert MaxText weights directly into target vLLM formats without requiring pre-initialized target model structures.
- Implements streaming parameter conversion to minimize host peak RSS during staging.
- Adds comprehensive unit tests in `tests/post_training/unit/weight_converter_test.py`.

## Verification
- `pytest tests/post_training/unit/weight_converter_test.py` (34/34 passed).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/integration/vllm/validate_converter.py [hunk 1]
```diff
@@ -256,43 +256,14 @@ def _compat_kv_cache_shape(*args, **kwargs):
     tpu_kv_cache = None
     orig_get_kv_cache_shape_with_mesh = None
 
-  def _compat_wsc(x, shardings):
+  from maxtext.integration.vllm.convert_utils import tunix_compat_context
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/validate_converter.py]: do we have to put the import statement here or is it AI agent being lazy?

File: src/maxtext/integration/vllm/weight_converter.py [hunk 2]
```diff
@@ -306,23 +294,42 @@ def __init__(
...
           allow_unused_source_keys=allow_unused_source_keys,
           debug=debug,
+          prefuse_moe_weights=prefuse_moe_weights,
+          target_dtype=target_dtype,
       )
       logging.info("WeightConverter: direct MaxText-to-MaxText mode (debug=%s).", debug)
     else:
+      if self.rules is None and config is not None:
+        for candidate_key in [
+            getattr(config, "model_name", ""),
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/weight_converter.py]: do we need getattr here?

File: src/maxtext/integration/vllm/weight_converter.py [hunk 3]
```diff
@@ -764,13 +852,7 @@ def _build_plan(
...
 
   def _validate_plan(self, src_flat, tgt_flat, unmatched, consumed) -> None:
-    """Fails loudly rather than leaving rollout weights at their dummy values.
-
-    vLLM boots with `load_format="dummy"`, so a target leaf we never write
-    keeps *random* weights. That produces a quietly wrong reward curve
-    instead of an error, which is far more expensive to debug than a crash
-    at startup.
-    """
+    """Fails loudly rather than leaving rollout weights at their dummy values."""
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/weight_converter.py]: Looks like JetSki removed a bunch of useful comments here and below.

File: src/maxtext/integration/vllm/weight_converter.py [hunk 4]
```diff
@@ -974,9 +1129,62 @@ def convert(self, src_pytree: Any, target_state: Any = None) -> Dict[str, Any]:
...
+
+      nested = traverse_util.unflatten_dict(piece_result)
+      del piece_result
+      yield jax.tree_util.tree_map(
+          lambda x: nnx.Param(x),
+          nested,
+      )
+
+    del src_flat
+    reclaim_host_memory()
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/weight_converter.py]: Is this really needed? I am under the impression that all the conversion code happens in TPU. Why would host memory saturate?

**Review by igorts-git — APPROVED** [review]

**Review by A9isha — APPROVED** [review]

**Review by hengtaoguo — APPROVED** [review]

File: src/maxtext/integration/vllm/weight_converter.py (hunk 3 above)
Comment (A9isha) [reply src/maxtext/integration/vllm/weight_converter.py]: bringing this back

File: src/maxtext/integration/vllm/weight_converter.py (hunk 2 above)
Comment (A9isha) [reply src/maxtext/integration/vllm/weight_converter.py]: removed it and updated the one unit test which would have failed since it missed model_name config

File: src/maxtext/integration/vllm/weight_converter.py (hunk 4 above)
Comment (A9isha) [reply src/maxtext/integration/vllm/weight_converter.py]: It is needed I think. While the actual tensor data resides on the TPU, manipulating the model state structure requires host memory.  Specifically, operations like `traverse_util.flatten_dict` and `unflatten_dict` allocate hundreds of thousands of short-lived Python tuples and strings to represent the flattened keys for massive models. 
While Python eventually garbage collects these objects, the underlying memory allocator (`glibc`) often holds onto the freed heap rather than returning it to the OS immediately. For huge models, this fragmentation causes the host memory footprint to balloon until the VM hits an Out-Of-Memory (OOM) error. `reclaim_host_memory()` explicitly triggers `malloc_trim(0)` to force the allocator to return that freed memory to the OS. 

I also added an inline comment explaining this in the code to avoid future confusion

File: src/maxtext/integration/vllm/weight_converter.py [hunk 5]
```diff
@@ -839,14 +917,102 @@ def _fuse_moe_bulk(self, wi_0, wi_1, tgt_val, key_path: str):
...
+    if isinstance(wi_0, jax.ShapeDtypeStruct):
+      fused_shape = (wi_0.shape[0], wi_0.shape[2], 2 * target_intermediate)
+      return tuple(jax.ShapeDtypeStruct(fused_shape, wi_0.dtype) for _ in range(wi_0.shape[self.scan_axis]))
 
-    The scanned source is cast, aligned and fused *once*; the per-layer
-    arrays are then read out of a single unstack. Every target in a group
-    shares a shape and sharding by construction, so the first one is a
-    sound stand-in for all of them.
-    """
+    tgt_shape = (wi_0.shape[0], wi_0.shape[2], 2 * target_intermediate)
```
[the remark this answers is not in the document]
Comment (A9isha) [reply src/maxtext/integration/vllm/weight_converter.py]: done

File: src/maxtext/integration/vllm/validate_converter.py (hunk 1 above)
Comment (A9isha) [reply src/maxtext/integration/vllm/validate_converter.py]: moved it back to the top of the file

File: src/maxtext/integration/vllm/weight_converter.py [hunk 6]
```diff
@@ -839,14 +917,102 @@ def _fuse_moe_bulk(self, wi_0, wi_1, tgt_val, key_path: str):
...
+          if self.padded_base_moe_mlp_dim > unrolled_shape[1]:
+            unrolled_shape[1] = self.padded_base_moe_mlp_dim
+        elif last_key in ("wi_0", "wi_1", "wi"):
+          if self.padded_base_moe_mlp_dim > unrolled_shape[-1]:
+            unrolled_shape[-1] = self.padded_base_moe_mlp_dim
+      return tuple(jax.ShapeDtypeStruct(tuple(unrolled_shape), val.dtype) for _ in range(val.shape[self.scan_axis]))
+
+    if last_key in MOE_MLP_WEIGHT_NAMES and self.padded_base_moe_mlp_dim is not None:
+      if last_key == "wo":
+        intermediate_axis = 2
```
[the remark this answers is not in the document]
Comment (A9isha) [reply src/maxtext/integration/vllm/weight_converter.py]: done
